invenio_rdm_records/collections/models.py

```diff
@@ -0,0 +1,136 @@
+# it under the terms of the MIT License; see LICENSE file for more details.
+"""Collections models."""
+
+from invenio_communities.communities.records.models import CommunityMetadata
+from invenio_db import db
+from invenio_records.models import Timestamp
+from sqlalchemy import UniqueConstraint
+from sqlalchemy_utils.types import UUIDType
+from traitlets import ClassBasedTraitType
+
+
+# CollectionTree Table
+class CollectionTree(db.Model, Timestamp):
+    """Collection tree model."""
+
+    __tablename__ = "collection_tree"
+
+    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
+    community_id = db.Column(
+        UUIDType,
+        db.ForeignKey(CommunityMetadata.id, ondelete="SET NULL"),
+        nullable=True,
+        index=True,
+    )
+    title = db.Column(db.String(255), nullable=False)
+    order = db.Column(db.Integer, nullable=True)
+    slug = db.Column(db.String(255), nullable=False, index=True)
+
+    # Relationship to Collection
+    collections = db.relationship("Collection", back_populates="collection_tree")
+    community = db.relationship(CommunityMetadata, backref="collection_trees")
+
+    # Unique constraint on slug and community_id. Slugs should be unique within a community.
+    UniqueConstraint(
+        "slug", "community_id", name="uq_collection_tree_slug_community_id"
+    )
+
+    @classmethod
+    def create(cls, title, slug, community_id=None, order=None):
+        """Create a new collection tree."""
+        with db.session.begin_nested():
+            collection_tree = cls(
+                title=title, slug=slug, community_id=community_id, order=order
+            )
+            db.session.add(collection_tree)
+        return collection_tree
+
+    @classmethod
+    def get(cls, id_):
+        """Get a collection tree by ID."""
+        return cls.query.get(id_)
+
+    @classmethod
+    def get_by_slug(cls, slug, community_id):
+        """Get a collection tree by slug."""
+        return cls.query.filter(
+            cls.slug == slug, cls.community_id == community_id
+        ).one_or_none()
+
+
+# Collection Table
+class Collection(db.Model, Timestamp):
+    """Collection model.
+
+    Indices:
+    - id
+    - collection_tree_id
+    - path
+    - slug
+    """
+
+    __tablename__ = "collections"
+
+    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
+    slug = db.Column(db.String(255), unique=False, nullable=False, index=True)
+    path = db.Column(db.Text, nullable=False, index=True)
+    tree_id = db.Column(db.Integer, db.ForeignKey("collection_tree.id"), nullable=False)
+    logo = db.Column(db.Text, nullable=True)
+    title = db.Column(db.String(255), nullable=False)
+    query_text = db.Column("query", db.Text, nullable=False)
+    order = db.Column(db.Integer, nullable=True)
+    # TODO index depth
+    depth = db.Column(
+        db.Integer, db.Computed("array_length(string_to_array(path, ','), 1) - 2")
```

Comment:
`depth` is computed server side. This funcion is PostgreSQL specific, however it can easily be moved to client-side if needed.